docs/discussion/concepts/overflow.md

```diff
@@ -0,0 +1,208 @@
+The key concept is the "smallest overflowing value".  For every combination of "conversion factor"
+and "type," there is some smallest starting-value that will overflow.  The simplest adaptive policy
+is to forbid conversions when that smallest value is "small enough to be scary".
+
+How small is "scary"?  Here are some considerations.
+
+- Once our values get over 1,000, we can consider switching to a larger SI-prefixed version of the
+  unit.  (For example, lengths over $1000\,\text{m}$ can be approximated in $\text{km}$.)  This
+  means that if a value as small as 1,000 would overflow --- so small that we haven't even _reached_
+  the next unit --- we should _definitely_ forbid the conversion.
+
+- On the other hand, we've found it useful to initialize, say, `QuantityI32<Hertz>` variables with
+  something like `mega(hertz)(500)`.  Thus, we'd like this operation to succeed (although it should
+  probably be near the border of what's allowed).
+
+Putting it all together, we settled on [a value threshold of 2'147][threshold].  If we can convert
+this value without overflow, then we permit the operation; otherwise, we don't.  We picked this
+value because it satisfies our above criteria nicely.  It will prevent operations that can't handle
+values of 1,000, but it still lets us use $\text{MHz}$ freely when storing $\text{Hz}$ quantities in
+`int32_t`.
+
+#### Plot: the Overflow Safety Surface
+
+This policy lends itself well to visualization.  For each integral type, there is some _biggest
+allowable conversion factor_ under this policy.  We can plot these factors for each of the common
+integral types (`int8_t`, `uint32_t`, and so on).  If we then "connect the dots", we get a boundary
+that separates allowed conversions from forbidden ones, permitting bigger conversions for bigger
+types. We call this abstract boundary the **"overflow safety surface"**, and it's the secret
+ingredient that lets Au users use a wide variety of integral types with confidence.
+
+![The overflow safety surface](../../assets/overflow-safety-surface.png)
+
+### Check every conversion at runtime
+
+While the overflow safety surface is a leap forward in safety and flexibility, it's still only
+a heuristic.  There will always be valid conversions which it forbids, and invalid ones which it
+permits.  On the latter point, note that adding an intermediate conversion can defeat the safety
+check: the overflow in `meters(10u).as(nano(meters))` would be caught, but the overflow in
+`meters(10u).as(milli(meters)).as(nano(meters))` would not.
+
+One way to _guarantee_ doing better is to check every conversion at runtime.  Some users may recoil
+at the idea of doing _runtime_ work in a units library, but it's easy to show that _this_ use case
+is innocuous. Consider: it's very hard to imagine a valid use case for performing unit conversions
+in a "hot loop".  Therefore, the extra runtime cost --- merely a few cycles at most --- won't
```

Comment:
Care must be taken to avoid "implicit" unit conversions inside a hot loop, no? For example your motivating example of evaluating `a>b`.

Reply:
True.  Tweaked "performing" to "needing to perform".  I didn't see an easy way to bring up the _implicit_ conversions _explicitly_ here without hurting the flow.

(But yeah, unintended implicit conversions could hurt performance.  I think the approach here would be something like what I used to hear at Google: "if you care about performance, and you don't have a benchmark, then you don't care about performance".  :grin:  I also got a question at CppCon 2023 about the possibility of making **all** unit conversions explicit --- which, although we're not planning to do it, _would_ solve this kind of problem.)